wagl/acquisition/landsat.py

```diff
@@ -144,6 +150,95 @@ def __init__(self, pathname, uri, acquisition_datetime, band_name='BAND 1',
         self._classification_type = 'U'
         self._international_designator = '99020A'
 
+        self._gap_mask = None
+
+    def _extract_gap_mask(self):
+        """
+        Extract the gap mask from within the tar.
+        The gap mask is a TIF that is then pushed through gzip,
+        horrible, and in order to read it, we need to unpack from
+        the tar, decompress using gzip, then write to disk so GDAL
+        can read it.
+        """
+        # mask files are contained in a sub-directory named 'gap-mask'
+        path = Path(self.uri.replace('!', '')[6:])
+        parts = path.name.split('_')
+
+        # non-thermal bands are named
+        # LE07_L1TP_092084_20110809_20161206_01_T1_B4.TIF
+        # thermal bands are named
+        # LE07_L1TP_092084_20110809_20161206_01_T1_B6_VCID_1.TIF
+        # so insert before index 7
+        # (not ideal, but USGS could change the whole convention anyway
+        parts.insert(7, 'GM')
+        mask_name = Path('gap_mask', '{}.gz'.format('_'.join(parts)))
+
+        # open tarfile
+        with tarfile.open(str(path.parent)) as tf:
+            mem = tf.getmember(str(mask_name))
+            fobj = tf.extractfile(mem)
+            with gzip.open(fobj) as gz:
+                with tempfile.TemporaryDirectory(suffix='.gap-mask') as tmpd:
+                    out_fname = Path(tmpd, 'gap-mask.tif')
+                    with open(out_fname, 'wb') as src:
+                        src.write(gz.read())
+
+                    # read gap mask into memory
+                    with rasterio.open(out_fname) as src:
+                        data = src.read(1)
+
+                    self._gap_mask = data == 0
+
+    def radiance_data(self, window=None, out_no_data=-999, esun=None):
+        """
+        This method overwrites the parent's method to handle a special
+        case for Landsat-7. The data supplied by USGS contains ~2pixel
+        interpolation for the SLC-OFF data. We need to read the
+        gap_mask file in order to mask out the interpolation fill.
+        Return the data as radiance in watts/(m^2*micrometre) with
+        interpolated pixels masked out.
+        """
+        # retrieve the gap mask if we haven't already done so
+        if self._gap_mask is None:
+            try:
+                self._extract_gap_mask()
```

Comment:
Just to get my head the logic in this block. If extraction of gap mask fails then the assumption is there is nothing to be masked?

Reply:
Yep.  GA's flavour of LPGS can't produce the gap mask files, and any Landsat-7 acquisitions prior to 2003-05-31 naturally won't have the gap mask files. So we simply handle both cases as nothing to be masked.
GA's flavour of LPGS won't be including the interpolation flag/switch that USGS includes.  In this way, our assumption of nothing to be masked will get around the fact that GA's flavour of LPGS can't produce the gap masks in the first place.